src/GafferOSL/OSLObject.cpp

```diff
@@ -385,3 +407,152 @@ ConstShadingEnginePtr OSLObject::shadingEngine( const Gaffer::Context *context )
+		// all the redundant recompiles are of shorter code.
+		outputCombineOSLCode()->codePlug()->setValue( "" );
+
+		outputCombineOSLCode()->parametersPlug()->clearChildren();
+
+		std::string code;
+
+		for( NameValuePlugIterator inputPlug( primitiveVariablesPlug() ); !inputPlug.done(); ++inputPlug )
+		{
+			std::string prefix = "";
+			BoolPlug* enabledPlug = (*inputPlug)->enabledPlug();
+			if( enabledPlug )
+			{
+				IntPlugPtr codeEnablePlug = new IntPlug( "enable" );
+				outputCombineOSLCode()->parametersPlug()->addChild( codeEnablePlug );
+				codeEnablePlug->setInput( enabledPlug );
+				prefix = "if( " + codeEnablePlug->getName().string() + " ) ";
+			}
+
+			PlugPtr valuePlug = (*inputPlug)->valuePlug<Plug>();
+
+			if( valuePlug->typeId() == ClosurePlug::staticTypeId() )
+			{
+				// Closures are a special case that doesn't need a wrapper function
+				ClosurePlugPtr codeClosurePlug = new ClosurePlug( "closureIn" );
+				outputCombineOSLCode()->parametersPlug()->addChild( codeClosurePlug );
+				codeClosurePlug->setInput( valuePlug );
+
+				code += prefix + "out += " + codeClosurePlug->getName().string() + ";\n";
+				continue;
+			}
+
+			std::string outFunction;
+			PlugPtr codeValuePlug;
+			const Gaffer::TypeId valueType = (Gaffer::TypeId)valuePlug->typeId();
+			switch( (int)valueType )
+			{
+				case FloatPlugTypeId :
+					codeValuePlug = new FloatPlug( "value" );
+					outFunction = "outFloat";
+					break;
+				case IntPlugTypeId :
+					codeValuePlug = new IntPlug( "value" );
+					outFunction = "outInt";
+					break;
+				case Color3fPlugTypeId :
+					codeValuePlug = new Color3fPlug( "value" );
+					outFunction = "outColor";
+					break;
+				case V3fPlugTypeId :
+					codeValuePlug = new V3fPlug( "value" );
+					{
+						V3fPlug *v3fPlug = runTimeCast<V3fPlug>( valuePlug.get() );
+						if( v3fPlug->interpretation() == GeometricData::Point )
+						{
+							outFunction = "outVector";
+						}
+						else if( v3fPlug->interpretation() == GeometricData::Normal )
+						{
+							outFunction = "outNormal";
+						}
+						else
+						{
+							outFunction = "outVector";
+						}
+					}
+					break;
+				case V2fPlugTypeId :
+					codeValuePlug = new V3fPlug( "value" );
+					outFunction = "outUV";
+					break;
+				case M44fPlugTypeId :
+					codeValuePlug = new M44fPlug( "value" );
+					outFunction = "outMatrix";
+					break;
+				case StringPlugTypeId :
+					codeValuePlug = new StringPlug( "value" );
+					outFunction = "outString";
+					break;
+			}
+
+			if( codeValuePlug )
+			{
+
+				StringPlugPtr codeNamePlug = new StringPlug( "name" );
+				outputCombineOSLCode()->parametersPlug()->addChild( codeNamePlug );
+				codeNamePlug->setInput( (*inputPlug)->namePlug() );
+
+				outputCombineOSLCode()->parametersPlug()->addChild( codeValuePlug );
+				if( (int)valueType == V2fPlugTypeId )
+				{
+					// For UVs, just connect up the first two components ( the third will be ignored by
```

Comment:
I mostly just wanted something not confusing for the UI, and figured that since we accept V3f -> V2f connections anyway, I might as well use that as the plug type.

Reply:
I can see why the V2f seems nice in the UI when looking at the OSLObject node in isolation, but I'm not convinced when you consider that its part of an ecosystem where we're not providing V2f anywhere...